### PR TITLE
FEATURE: Classification Visualiser Tool

### PR DESCRIPTION
**Pull request by zacbrannelly**
From #133 

**This iteration todos:**
- [x] Order confusion matrix by most popular class to least popular class
- [x] Output file format in HTML or JSON. Always print to the screen.
- [x] Visualisation function should be different from function the output metrics
- [x] Wrap in a Visualiser interface for use in Surround
- [x] Support multiple ground truth and prediction columns
- [x] Add flag to output file with only the filtered failed cases. Output the same file format as the input data format. True by default.
- [x] Rename module to visualise_classifier.py
- [x] Integrate into the Surround CLI
- [x] Add normalize flag
- [x] Separate reports for ground truth / predict pairs
- [x] Output HTML and JSON reports by default
- [x] Remove matplotlib dependency, use https://bl.ocks.org/arpitnarechania/dbf03d8ef7fffa446379d59db6354bac instead

**Future work:**
- Add a flag to set probability thresholds
- Add flag to provide more human readable explanations of metrics

Example Output:
```
$ surround viz --help
usage: surround visualise [-h] [-g GROUND_TRUTH] [-p PREDICTIONS]
                          [-s SEPARATER] [--header HEADER]
                          [-o OUTPUT_DIRECTORY] [-i OUTPUT_INCORRECT] [-n]
                          [-no] [-jo] [-ho]
                          data_file

positional arguments:
  data_file             CSV file with ground truth and predicted labels

optional arguments:
  -h, --help            show this help message and exit
  -g GROUND_TRUTH, --ground-truth GROUND_TRUTH
                        Comma separated list of columns containing ground
                        truth values (default: ground_truth)
  -p PREDICTIONS, --predictions PREDICTIONS
                        Comma separated list of columns containing predictions
                        (default: predictions)
  -s SEPARATER, --separater SEPARATER
                        File separator (default: ',')
  --header HEADER       Row of the file to use as a header (default: 0)
  -o OUTPUT_DIRECTORY, --output-directory OUTPUT_DIRECTORY
                        Path to directory to export JSON/HTML results to
  -i OUTPUT_INCORRECT, --output-incorrect OUTPUT_INCORRECT
                        Output file containing incorrect records (default:
                        true)
  -n, --normalize       Show the normalized confusion matrix in reports
  -no, --no-output      Don't output reports to JSON/HTML files
  -jo, --json-only      Output visualisations to JSON files
  -ho, --html-only      Output visualisations to HTML files
```
```
$ surround viz -g g1 -p p1 surround/test.csv
Results for columns g1 & p1
Ground truth column label: g1
Predict column label: p1
============[Clasification Report]===================
Overall Metrics:
              |precision |recall    |f1-score
macro avg     |0.67      |0.39      |0.49
weighted avg  |1.00      |0.60      |0.75

Accuracy: 0.6
Cohen Kappa: 0.4118
==========================
Metrics per category:
        |accuracy  |precision |recall    |f1-score  |support
Cat     |0.00      |0.00      |0.00      |0.00      |0.00
Dog     |0.67      |1.00      |0.67      |0.80      |3.00
Mouse   |0.50      |1.00      |0.50      |0.67      |2.00

==========================
Confusion Matrix:
       Cat   Dog  Mouse
  Cat|  0  |  0  |  0  |
  Dog|  1  |  2  |  0  |
Mouse|  1  |  0  |  1  |
============[End of visualisation]===================

Exported incorrect values to file: D:\Git Repos\surround\incorrect_records.csv
Exported a JSON report to file: D:\Git Repos\surround\report_g1_p1.json
Exported a HTML report to file: D:\Git Repos\surround\report_g1_p1.html
```

Example JSON output:
```
{
    "input_file": "test.csv",
    "ground_truth_label": "g1",
    "predict_label": "p1",
    "results": {
        "report": {
            "Cat": {
                "precision": 0.0,
                "recall": 0.0,
                "f1-score": 0.0,
                "support": 0,
                "accuracy": 0.0
            },
            "Dog": {
                "precision": 1.0,
                "recall": 0.6667,
                "f1-score": 0.8,
                "support": 3,
                "accuracy": 0.6667
            },
            "Mouse": {
                "precision": 1.0,
                "recall": 0.5,
                "f1-score": 0.6667,
                "support": 2,
                "accuracy": 0.5
            },
            "accuracy": 0.6,
            "macro avg": {
                "precision": 0.6667,
                "recall": 0.3889,
                "f1-score": 0.4889,
                "support": 5
            },
            "weighted avg": {
                "precision": 1.0,
                "recall": 0.6,
                "f1-score": 0.7467,
                "support": 5
            }
        },
        "confusion_matrix": [
            [
                2,
                0,
                1
            ],
            [
                0,
                1,
                1
            ],
            [
                0,
                0,
                0
            ]
        ],
        "normalized_confusion_matrix": [
            [
                0.6667,
                0.0,
                0.3333
            ],
            [
                0.0,
                0.5,
                0.5
            ],
            [
                0.0,
                0.0,
                0.0
            ]
        ],
        "accuracy_score": 0.6,
        "cohen_kappa_score": 0.4118,
        "classes": [
            "Dog",
            "Mouse",
            "Cat"
        ]
    }
}
```

### CODE REVIEW

File: surround/cli.py [hunk 1]
```diff
@@ -456,9 +459,11 @@ def execute_cli():
     remote_cli.add_push_parser(sub_parser)
     remote_cli.add_list_parser(sub_parser)
 
+    visualise_parser = sub_parser.add_parser('visualise', parents=[visualise_cli.get_visualise_parser()], add_help=False, help="Visualise results of a pipeline")
+
     # Check for valid sub commands as 'add_subparsers' in Python < 3.7
     # is missing the 'required' keyword
-    tools = ["init", "lint", "run", "remote", "add", "pull", "push", "list"]
+    tools = ["init", "lint", "run", "remote", "add", "pull", "push", "list", "visualise"]
```
Comment (ScottyB) [inline surround/cli.py]: `visualise` is too long, what about `viz`?

File: surround/visualise/cli.py [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+    parser.add_argument("-p", "--predictions", help="Comma separated list of columns containing predictions (default: predictions)", default="predictions")
+    parser.add_argument("-s", "--separater", help="File separator (default: ',')", default=",")
+    parser.add_argument("--header", type=int, help="Row of the file to use as a header (default: 0)", default=0)
+    parser.add_argument('-o', "--output-directory", help="Path to directory to export JSON/HTML results to", default=os.path.abspath("."), type=lambda x: is_valid_directory(parser, x))
+    parser.add_argument('-i', '--output-incorrect', help="Output file containing incorrect records (default: true)", default=True, type=str2bool)
+    parser.add_argument('-n', '--normalize', action="store_true", help="Show the normalized confusion matrix in reports")
+    parser.add_argument('-no', '--no-output', action="store_true", help="Don't output reports to JSON/HTML files")
+    parser.add_argument('-jo', '--json-only', action="store_true", help="Output visualisations to JSON files")
```
Comment (ScottyB) [inline surround/visualise/cli.py]: I don't think we should generate the JSON or HTML files by default.

File: surround/cli.py (hunk 1 above)
Comment (zacbrannelly) [reply surround/cli.py]: Yeah sure 👍

**Comment by zacbrannelly** [comment]
@ScottyB Should be able to use tab as a separator and no longer crash when the file cannot be read.

File: requirements.txt [hunk 3]
```diff
@@ -2,5 +2,4 @@ pyaml==17.12.1
 doit==0.31.1
 pandas==0.24.2
 scikit-learn==0.21.2
-matplotlib==3.0.2
```
Comment (luisfciencias) [inline requirements.txt]: 😿

**Comment by ScottyB** [comment]
@Nikhii can you please test this PR using the output from IOOF? We should be able to use this tool for the metrics if we have each category on a new line in the file as discussed today.

**Comment by Nikhii** [comment]
@ScottyB I have tested this PR.Looks fine-
Generates three files -
1)HTML
2)JSON
3)Failed predictions

These are tested on the 20000 (single labelled IOOF *.eml files) and they work fine.
For IOOF purposed i might have to add some of the multi-label metrics in.

**Pull request merged**